plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -263,11 +284,33 @@ constant application machinery over a list of 'CekValue's and turning the cause
 failure into a 'Term', apart from the straightforward generalization of the error type.
 -}
 
+{- Note [Being generic over @term@ in 'CekM']
+We have a @term@-generic version of 'CekM' called 'CekCarryingM', which itself requires a
+@term@-generic version of 'CekEvaluationException' called 'CekEvaluationExceptionCarrying'.
+This enables us to implement 'MonadError' instances that allow for throwing errors in both the CEK
+machine and the builtin application machinery (as Note [Being generic over 'term' in errors]
+explains, those are different kinds of errors) and otherwise we'd have to have more plumbing between
+these two and turning the builtin application machinery's 'MonadError' constraint into an explicit
+'Either' just to dispatch on it in the CEK machine and rethrow the error non-purely  (we used to
+have that) has to cost some. Plus, being @term@-generic also enables us to throw actual exceptions
```

Comment:
@michaelpj you're focusing on the wrong thing. Focus on these four lines. We used to have that dispatching on an explicit `Either` result whenever we do any unlifting, now we don't. I haven't checked whether it improves performance much or just a little bit, but it just looks like a very reasonable thing to do. Particularly given that it also allows us to drop a lot of code duplication and those pesky type applications of impurely throwing functions.

Reply:
Ok, I did that.